File: tools/release/verify_cli_release.py
#!/usr/bin/env python3
"""Verify the complete set of CWE-193 CLI release assets."""

import argparse
import hashlib
import json
from pathlib import Path
import tarfile
import zipfile


ASSETS = (
    "caudex-workout-cli-0.1.0-x86_64-linux-gnu.tar.gz",
    "caudex-workout-cli-0.1.0-aarch64-linux-gnu.tar.gz",
    "caudex-workout-cli-0.1.0-x86_64-macos.tar.gz",
    "caudex-workout-cli-0.1.0-aarch64-macos.tar.gz",
    "caudex-workout-cli-0.1.0-x86_64-windows-gnu.zip",
)
REQUIRED_COMMON = {"LICENSE", "NOTICE", "README.md", "build-metadata.json"}


def checksum(path: Path) -> str:
    digest = hashlib.sha256()
    with path.open("rb") as source:
        for chunk in iter(lambda: source.read(1024 * 1024), b""):
            digest.update(chunk)
    return digest.hexdigest()


def archive_entries(path: Path):
    if path.name.endswith(".zip"):
        with zipfile.ZipFile(path) as archive:
            return archive.namelist()
    with tarfile.open(path, "r:gz") as archive:
        return archive.getnames()


def verify_archive(directory: Path, name: str) -> None:
    path = directory / name
    if not path.is_file():
        raise SystemExit(f"missing release archive: {name}")
    stem = name.removesuffix(".tar.gz").removesuffix(".zip")
    entries = archive_entries(path)
    prefix = stem + "/"
    if any(not entry.startswith(prefix) or ".." in Path(entry).parts for entry in entries):
        raise SystemExit(f"unsafe archive path in {name}")
    files = {entry.removeprefix(prefix) for entry in entries if not entry.endswith("/")}
    required = REQUIRED_COMMON | ({"caudex.exe", "sqlite3.dll"} if name.endswith(".zip") else {"caudex"})
    if files != required:
        raise SystemExit(f"unexpected files in {name}: {sorted(files)}")

    if name.endswith(".zip"):
        import zipfile

        with zipfile.ZipFile(path) as archive:
            metadata = json.loads(archive.read(f"{stem}/build-metadata.json"))
    else:
        with tarfile.open(path, "r:gz") as archive:
            metadata = json.loads(archive.extractfile(f"{stem}/build-metadata.json").read())
    if metadata.get("version") != "0.1.0" or metadata.get("optimize") != "ReleaseSafe":
        raise SystemExit(f"invalid build metadata in {name}")
    target = name.removesuffix(".tar.gz").removesuffix(".zip").removeprefix("caudex-workout-cli-0.1.0-")
    if metadata.get("target") != target:
        raise SystemExit(f"target metadata does not match {name}")
    if metadata.get("archiveStem") != stem or metadata.get("platformCodeSigned") is not False:
        raise SystemExit(f"invalid provenance metadata in {name}")
    sqlite = metadata.get("sqlite", {})
    if not all(sqlite.get(key) for key in ("version", "linkage", "source")):
        raise SystemExit(f"incomplete SQLite metadata in {name}")
    if len(sqlite.get("checksum", {}).get("value", "")) != 64:
        raise SystemExit(f"missing SQLite checksum in {name}")


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("directory", type=Path)
    args = parser.parse_args()
    directory = args.directory
    sums = directory / "SHA256SUMS"
    if not sums.is_file():
        raise SystemExit("missing SHA256SUMS")
    expected = {}
    for line in sums.read_text().splitlines():
        digest, name = line.split(maxsplit=1)
        expected[name.removeprefix("*")] = digest
    if set(expected) != set(ASSETS):
        raise SystemExit(f"SHA256SUMS does not cover exactly the release archives: {sorted(expected)}")
    for name in ASSETS:
        verify_archive(directory, name)
        if checksum(directory / name) != expected[name]:
            raise SystemExit(f"checksum mismatch: {name}")
    print("verified CLI release archives, metadata, and SHA256SUMS")


if __name__ == "__main__":
    main()

File: tools/release/verify-cli-release.mjs
#!/usr/bin/env node
import fs from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { archiveEntries, sha256 } from "./archive-utils.mjs";

export async function verifyCliRelease(directory, version, options = {}) {
  const assets = [
    "x86_64-linux-gnu", "aarch64-linux-gnu", "aarch64-macos", "x86_64-windows-gnu",
  ].map((target) => `caudex-workout-cli-${version}-${target}${target.endsWith("windows-gnu") ? ".zip" : ".tar.gz"}`);
  const suppliedChecksums = options.checksums;
  const expected = suppliedChecksums ?? parseChecksums(
    await fs.readFile(path.join(directory, "SHA256SUMS"), "utf8"),
  );
  if (
    (!(suppliedChecksums instanceof Map) && expected.size !== assets.length) ||
    assets.some((name) => !expected.has(name))
  ) {
    throw new Error("SHA256SUMS does not cover exactly the release CLI archives");
  }
  for (const name of assets) {
    const archive = path.join(directory, name);
    if (await sha256(archive) !== expected.get(name)) throw new Error(`checksum mismatch: ${name}`);
    verifyArchive(await archiveEntries(archive), name, version);
  }
}

if (process.argv[1] && path.resolve(process.argv[1]) === fileURLToPath(import.meta.url)) {
  const directory = path.resolve(process.argv[2] ?? "");
  const versionIndex = process.argv.indexOf("--version");
  const version = versionIndex >= 0 ? process.argv[versionIndex + 1] : undefined;
  if (!directory || !version) throw new Error("usage: node tools/release/verify-cli-release.mjs DIRECTORY --version VERSION");
  await verifyCliRelease(directory, version);
  console.log("verified CLI release archives, metadata, and SHA256SUMS");
}

function parseChecksums(contents) {
  const checksums = new Map();
  for (const line of contents.split(/\r?\n/).filter(Boolean)) {
    const match = line.match(/^([0-9a-f]{64})\s+\*?([^\s]+)$/i);
    if (!match || path.basename(match[2]) !== match[2]) throw new Error(`unsafe checksum path: ${line}`);
    checksums.set(match[2], match[1]);
  }
  return checksums;
}

function verifyArchive(entries, name, version) {
  const stem = name.replace(/\.tar\.gz$|\.zip$/, "");
  const prefix = `${stem}/`;
  if (entries.some(({ name: entry }) => !entry.startsWith(prefix) || entry.includes("\\") || entry.split("/").includes(".."))) throw new Error(`unsafe archive path in ${name}`);
  const files = new Set(entries.filter(({ name: entry }) => !entry.endsWith("/")).map(({ name: entry }) => entry.slice(prefix.length)));
  const required = new Set(["LICENSE", "NOTICE", "README.md", "build-metadata.json", name.endsWith(".zip") ? "caudex.exe" : "caudex"]);
  if (files.size !== required.size || [...required].some((entry) => !files.has(entry))) throw new Error(`unexpected files in ${name}: ${[...files].sort().join(", ")}`);
  const metadataEntry = entries.find(({ name: entry }) => entry === `${stem}/build-metadata.json`);
  const metadata = JSON.parse(metadataEntry.data.toString("utf8"));
  const target = stem.slice(`caudex-workout-cli-${version}-`.length);
  if (metadata.version !== version || metadata.target !== target || metadata.optimize !== "ReleaseSafe" || metadata.archiveStem !== stem || metadata.platformCodeSigned !== false) throw new Error(`invalid build metadata in ${name}`);
  const sqlite = metadata.sqlite ?? {};
  if (!["version", "linkage", "source"].every((key) => sqlite[key]) || sqlite.checksum?.algorithm !== "sha256" || !/^[0-9a-f]{64}$/i.test(sqlite.checksum?.value ?? "")) throw new Error(`incomplete SQLite metadata in ${name}`);
}
